hedera-node/docs/design/services/token-service/airdrops/airdrop-pending-state.md

```diff
@@ -0,0 +1,202 @@
+   For a specific `PendingAirdropId`, we would get the corresponding `PendingAirdropValue`. For NFT airdrops we don't have a corresponding `PendingAirdropValue` and the information for them would be encapsulated in the `PendingAirdropId` itself.
+2. Fetching all PendingAirdropIds, so that we can iterate over them and perform business logic (e.g. in case of sender or token expirations)
+3. Handling a new airdrop in state. We can have several cases:
+   - Only a single fungible or non-fungible airdrop for a specific token defined - we just put a new entry in the map
+   - An airdrop for an existing `PendingAirdropId` is already defined - we should get the existing `PendingAirdropValue` for this `PendingAirdropId` and modify it with the aggregated new fungible amount
+4. Handling an airdrop `cancel` in state - we remove the `PendingAirdropId` key for the airdrop we want to cancel
+5. Handling an airdrop `claim` in state - we remove the `PendingAirdropId` key for the airdrop we want to claim
+
+## Acceptance tests
+
+The airdrop state correctness would be validated by acceptance tests performed by the `TokenAirdropTransaction`, `TokenCancelAirdrop` and `TokenClaimAirdrop`.
+
+The airdrop state itself could be covered by a set of unit tests, covering different corner cases, as well as positive and negative scenarios.
+
+The main behaviour that should be validated for the airdrop state and its storage is the following:
+
+Note: For all cases the recipient account shouldn’t be associated with the airdropped token.
+
+### State management on airdrop (populate pending airdrop state only)
+
+1. Airdrop a fungible token with a negative amount to Alice → the airdrop should fail and nothing should be saved in the pending state
+2. Airdrop a fungible token with a `Long.MAX_VALUE` amount to Alice and then make a second airdrop with the same fungible token and amount with a value of 1 again to Alice → the second airdrop attempt should fail and it’s data shouldn’t be saved in the pending state
+3. Airdrop an NFT with a negative serial number to Alice → the airdrop should fail and nothing should be saved in the pending state
+4. Airdrop an NFT with a serial number 1 to Alice and then perform another airdrop with the same token and a serial number 2 again to Alice → the airdrop should succeed and all of the airdrop data should be saved in the pending state
+5. Airdrop a fungible token with amount X to Alice and the same fungible token with amount Y to Bob → the airdrops should succeed and all airdrop data should be saved in the pending state
+6. Airdrop a fungible token with amount X to Alice and the same fungible token with amount Y again to Alice → the airdrops should succeed and the airdrop data should be aggregated having an amount of X + Y and should be saved in the pending state
```

Comment:
What is the difference between 5. and 6. Is 5. meant to be for NFTs?

Reply:
Both are connected to fungible tokens. The difference is that in 6. we have the same recipient for separate airdrops and the end result would be aggregation of the two.

Reply:
Ah, now I see it. Thanks!